File: src/main/java/jugglinglab/jml/JMLEvent.java
//
// JMLEvent.java
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml;

import java.io.*;
import java.util.*;
import jugglinglab.util.*;

public class JMLEvent {
  static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

  protected double x, y, z;  // coordinates in local frame
  protected double gx, gy, gz;  // coordinates in global frame
  protected boolean globalvalid;  // global coordinates need to be recalced?
  protected double t;
  protected int juggler;
  protected int hand;
  protected ArrayList<JMLTransition> transitions;
  protected int delay;
  protected int delayunits;
  protected Permutation pathpermfrommaster;
  protected JMLEvent master;  // null if this is a master event
  public boolean calcpos;

  protected JMLEvent prev, next;  // for doubly-linked event list

  public JMLEvent() {
    calcpos = false;
    transitions = new ArrayList<>();
  }

  public Coordinate getLocalCoordinate() {
    return new Coordinate(x, y, z);
  }

  public void setLocalCoordinate(Coordinate c) {
    x = c.x;
    y = c.y;
    z = c.z;
    globalvalid = false;
  }

  public Coordinate getGlobalCoordinate() {
    return (globalvalid ? new Coordinate(gx, gy, gz) : null);
  }

  public void setGlobalCoordinate(Coordinate c) {
    gx = c.x;
    gy = c.y;
    gz = c.z;
    globalvalid = true;
  }

  public double getT() {
    return t;
  }

  public void setT(double time) {
    t = time;
  }

  public int getHand() {
    return hand;
  }

  public void setHand(String strhand) throws JuggleExceptionUser {
    int index = strhand.indexOf(":");

    if (index == -1) {
      juggler = 1;
      if (strhand.equalsIgnoreCase("left")) {
        hand = HandLink.LEFT_HAND;
      } else if (strhand.equalsIgnoreCase("right")) {
        hand = HandLink.RIGHT_HAND;
      } else {
        throw new JuggleExceptionUser(
            errorstrings.getString("Error_hand_name") + " '" + strhand + "'");
      }
    } else {
      juggler = Integer.parseInt(strhand.substring(0, index));
      String substr = strhand.substring(index + 1);
      if (substr.equalsIgnoreCase("left")) {
        hand = HandLink.LEFT_HAND;
      } else if (substr.equalsIgnoreCase("right")) {
        hand = HandLink.RIGHT_HAND;
      } else {
        throw new JuggleExceptionUser(
            errorstrings.getString("Error_hand_name") + " '" + strhand + "'");
      }
    }
  }

  public void setHand(int j, int h) {
    juggler = j;
    hand = h;  // HandLink.LEFT_HAND or HandLink.RIGHT_HAND
  }

  public int getJuggler() {
    return juggler;
  }

  public int getNumberOfTransitions() {
    return transitions.size();
  }

  public JMLTransition getTransition(int index) {
    return transitions.get(index);
  }

  public Collection<JMLTransition> transitions() {
    return transitions;
  }

  public void addTransition(JMLTransition trans) {
    transitions.add(trans);
  }

  public void removeTransition(int index) {
    transitions.remove(index);
  }

  public void removeTransition(JMLTransition trans) {
    transitions.remove(trans);
  }

  public boolean isMaster() {
    return (master == null);
  }

  public JMLEvent getMaster() {
    return master;
  }

  public void setMaster(JMLEvent master) {
    this.master = master;
  }

  public JMLEvent getPrevious() {
    return prev;
  }

  public void setPrevious(JMLEvent prev) {
    this.prev = prev;
  }

  public JMLEvent getNext() {
    return next;
  }

  public void setNext(JMLEvent next) {
    this.next = next;
  }

  public JMLEvent getPreviousForHand() {
    JMLEvent ev = getPrevious();

    while (ev != null) {
      if (ev.getJuggler() == getJuggler() && ev.getHand() == getHand()) {
        return ev;
      }
      ev = ev.getPrevious();
    }
    return null;
  }

  public JMLEvent getNextForHand() {
    JMLEvent ev = getNext();

    while (ev != null) {
      if (ev.getJuggler() == getJuggler() && ev.getHand() == getHand()) {
        return ev;
      }
      ev = ev.getNext();
    }
    return null;
  }

  public Permutation getPathPermFromMaster() {
    return pathpermfrommaster;
  }

  public void setPathPermFromMaster(Permutation p) {
    pathpermfrommaster = p;
  }

  public boolean isDelayOf(JMLEvent ev2) {
    JMLEvent mast1 = (getMaster() == null ? this : getMaster());
    JMLEvent mast2 = (ev2.getMaster() == null ? ev2 : ev2.getMaster());

    if (mast1 != mast2) {
      return false;
    }
    if (getJuggler() != ev2.getJuggler() || getHand() != ev2.getHand()) {
      return false;
    }

    int totaldelay = delay - ev2.delay;
    if (totaldelay < 0) {
      totaldelay = -totaldelay;
    }
    return (totaldelay % delayunits) == 0;
  }

  public boolean hasSameMasterAs(JMLEvent ev2) {
    JMLEvent mast1 = (getMaster() == null ? this : getMaster());
    JMLEvent mast2 = (ev2.getMaster() == null ? ev2 : ev2.getMaster());

    return (mast1 == mast2);
  }

  public JMLTransition getPathTransition(int path, int transtype) {
    for (JMLTransition tr : transitions) {
      if (tr.getPath() == path) {
        if (transtype == JMLTransition.TRANS_ANY || transtype == tr.getType()) {
          return tr;
        }
      }
    }
    return null;
  }

  public boolean hasThrow() {
    for (JMLTransition tr : transitions()) {
      if (tr.getType() == JMLTransition.TRANS_THROW) {
        return true;
      }
    }
    return false;
  }

  public boolean hasThrowOrCatch() {
    for (JMLTransition tr : transitions()) {
      int type = tr.getType();

      if (type == JMLTransition.TRANS_THROW
          || type == JMLTransition.TRANS_CATCH
          || type == JMLTransition.TRANS_SOFTCATCH
          || type == JMLTransition.TRANS_GRABCATCH) {
        return true;
      }
    }
    return false;
  }

  // Return true if the event contains a throw transition to another juggler.
  //
  // Note this will only work after pattern layout.

  public boolean hasPassingThrow() {
    for (JMLTransition tr : transitions()) {
      if (tr.getType() != JMLTransition.TRANS_THROW) {
        continue;
      }

      PathLink pl = tr.getOutgoingPathLink();
      if (pl == null || pl.getEndEvent() == null) {
        continue;
      }
      if (pl.getEndEvent().getJuggler() != getJuggler()) {
        return true;
      }
    }
    return false;
  }

  // Return true if the event contains a catch transition from another juggler.
  //
  // Note this will only work after pattern layout.

  public boolean hasPassingCatch() {
    for (JMLTransition tr : transitions()) {
      if (tr.getType() != JMLTransition.TRANS_CATCH
          && tr.getType() != JMLTransition.TRANS_SOFTCATCH
          && tr.getType() != JMLTransition.TRANS_GRABCATCH) {
        continue;
      }

      PathLink pl = tr.getIncomingPathLink();
      if (pl == null || pl.getStartEvent() == null) {
        continue;
      }
      if (pl.getStartEvent().getJuggler() != getJuggler()) {
        return true;
      }
    }
    return false;
  }

  // Note this will only work after pattern layout.

  public boolean hasPassingTransition() {
    return (hasPassingThrow() || hasPassingCatch());
  }

  public JMLEvent duplicate(int delay, int delayunits) {
    JMLEvent dup = new JMLEvent();
    dup.setLocalCoordinate(getLocalCoordinate());
    dup.setT(getT());
    dup.setHand(getJuggler(), getHand());
    dup.delay = delay;
    dup.delayunits = delayunits;
    dup.calcpos = calcpos;

    for (JMLTransition tr : transitions()) {
      dup.addTransition(tr.duplicate());
    }

    dup.setMaster(isMaster() ? this : master);
    return dup;
  }

  // For locating a particular event in a pattern.

  public int getHashCode() {
    Coordinate c = getLocalCoordinate();
    String s =
        "<event x=\""
            + JLFunc.toStringRounded(c.x, 4)
            + "\" y=\""
            + JLFunc.toStringRounded(c.y, 4)
            + "\" z=\""
            + JLFunc.toStringRounded(c.z, 4)
            + "\" t=\""
            + JLFunc.toStringRounded(getT(), 4)
            + "\" hand=\""
            + getJuggler()
            + ":"
            + (getHand() == HandLink.LEFT_HAND ? "left" : "right")
            + "\">";
    return s.hashCode();
  }

  //----------------------------------------------------------------------------
  // Reader/writer methods
  //----------------------------------------------------------------------------

  public void readJML(JMLNode current, String jmlvers, int njugglers, int npaths)
      throws JuggleExceptionUser {
    JMLAttributes at = current.getAttributes();
    double tempx = 0, tempy = 0, tempz = 0, tempt = 0;
    String handstr = null;

    try {
      for (int i = 0; i < at.getNumberOfAttributes(); i++) {
        // System.out.println("att. "+i+" = "+at.getAttributeValue(i));
        if (at.getAttributeName(i).equalsIgnoreCase("x")) {
          tempx = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("y")) {
          tempy = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("z")) {
          tempz = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("t")) {
          tempt = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("hand")) {
          handstr = at.getAttributeValue(i);
        }
      }
    } catch (NumberFormatException nfe) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_event_coordinate"));
    }

    // JML version 1.0 used a different coordinate system -- convert
    if (jmlvers.equals("1.0")) {
      double temp = tempy;
      tempy = tempz;
      tempz = temp;
    }

    setLocalCoordinate(new Coordinate(tempx, tempy, tempz));
    setT(tempt);
    if (handstr == null) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_unspecified_hand"));
    }
    setHand(handstr);
    if (juggler > njugglers || juggler < 1) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_juggler_out_of_range"));
    }

    // process current event node children
    for (int i = 0; i < current.getNumberOfChildren(); i++) {
      JMLNode child = current.getChildNode(i);
      String nodetype = child.nodeType;
      at = child.getAttributes();
      String path = null, transtype = null, mod = null;

      for (int j = 0; j < at.getNumberOfAttributes(); j++) {
        String value = at.getAttributeValue(j);
        if (at.getAttributeName(j).equalsIgnoreCase("path")) {
          path = value;
        } else if (at.getAttributeName(j).equalsIgnoreCase("type")) {
          transtype = value;
        } else if (at.getAttributeName(j).equalsIgnoreCase("mod")) {
          mod = value;
        }
      }

      if (path == null) {
        throw new JuggleExceptionUser(errorstrings.getString("Error_no_path"));
      }

      int pnum = Integer.parseInt(path);
      if (pnum > npaths || pnum < 1) {
        throw new JuggleExceptionUser(errorstrings.getString("Error_path_out_of_range"));
      }

      if (nodetype.equalsIgnoreCase("throw")) {
        addTransition(new JMLTransition(JMLTransition.TRANS_THROW, pnum, transtype, mod));
      } else if (nodetype.equalsIgnoreCase("catch") && transtype.equalsIgnoreCase("soft")) {
        addTransition(new JMLTransition(JMLTransition.TRANS_SOFTCATCH, pnum, null, null));
      } else if (nodetype.equalsIgnoreCase("catch") && transtype.equalsIgnoreCase("grab")) {
        addTransition(new JMLTransition(JMLTransition.TRANS_GRABCATCH, pnum, null, null));
      } else if (nodetype.equalsIgnoreCase("catch")) {
        addTransition(new JMLTransition(JMLTransition.TRANS_CATCH, pnum, null, null));
      } else if (nodetype.equalsIgnoreCase("holding")) {
        addTransition(new JMLTransition(JMLTransition.TRANS_HOLDING, pnum, null, null));
      }

      if (child.getNumberOfChildren() != 0) {
        throw new JuggleExceptionUser(errorstrings.getString("Error_event_subtag"));
      }
    }
  }

  public void writeJML(PrintWriter wr) throws IOException {
    Coordinate c = getLocalCoordinate();
    wr.println(
        "<event x=\""
            + JLFunc.toStringRounded(c.x, 4)
            + "\" y=\""
            + JLFunc.toStringRounded(c.y, 4)
            + "\" z=\""
            + JLFunc.toStringRounded(c.z, 4)
            + "\" t=\""
            + JLFunc.toStringRounded(getT(), 4)
            + "\" hand=\""
            + getJuggler()
            + ":"
            + (getHand() == HandLink.LEFT_HAND ? "left" : "right")
            + "\">");
    for (JMLTransition tr : transitions()) {
      tr.writeJML(wr);
    }
    wr.println("</event>");
  }

  // java.lang.Object methods

  @Override
  public String toString() {
    StringWriter sw = new StringWriter();
    try {
      writeJML(new PrintWriter(sw));
    } catch (IOException ioe) {
    }

    return sw.toString();
  }
}

File: composeApp/src/commonMain/kotlin/jugglinglab/jml/JmlEvent.kt
//
// JmlEvent.kt
//
// Copyright 2002-2026 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml

import jugglinglab.composeapp.generated.resources.*
import jugglinglab.util.Coordinate
import jugglinglab.util.jlToStringRounded
import jugglinglab.util.jlParseFiniteDouble
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.jlGetStringResource

data class JmlEvent(
    val x: Double = 0.0,
    val y: Double = 0.0,
    val z: Double = 0.0,
    val t: Double = 0.0,
    val juggler: Int = 1,
    val hand: Int = 0,
    val transitions: List<JmlTransition> = emptyList()
) : Comparable<JmlEvent> {
    val localCoordinate: Coordinate
        get() = Coordinate(x, y, z)

    val truncatedTime: Double by lazy {
        jlToStringRounded(t, 4).toDouble()
    }

    fun getPathTransition(path: Int, transType: Int): JmlTransition? {
        return transitions.firstOrNull {
            it.path == path && (transType == JmlTransition.TRANS_ANY || transType == it.type)
        }
    }

    val hasThrow: Boolean by lazy {
        transitions.any {
            it.type == JmlTransition.TRANS_THROW
        }
    }

    val hasThrowOrCatch: Boolean by lazy {
        transitions.any {
            when (it.type) {
                JmlTransition.TRANS_THROW,
                JmlTransition.TRANS_CATCH,
                JmlTransition.TRANS_SOFTCATCH,
                JmlTransition.TRANS_GRABCATCH -> true

                else -> false
            }
        }
    }

    fun writeJml(wr: Appendable, startTagOnly: Boolean = false) {
        wr.append("<event x=\"${jlToStringRounded(x, 4)}\"")
        wr.append(" y=\"${jlToStringRounded(y, 4)}\"")
        wr.append(" z=\"${jlToStringRounded(z, 4)}\"")
        wr.append(" t=\"${jlToStringRounded(t, 4)}\"")
        wr.append(" hand=\"$juggler:")
        wr.append(if (hand == LEFT_HAND) "left" else "right")
        wr.append("\">\n")
        if (!startTagOnly) {
            transitions.forEach { it.writeJml(wr) }
            wr.append("</event>\n")
        }
    }

    private val cachedToString: String by lazy {
        val sb = StringBuilder()
        writeJml(sb)
        sb.toString()
    }

    override fun toString(): String = cachedToString

    // Event hash code for locating a particular event in a pattern.

    val jlHashCode: Int by lazy {
        val sb = StringBuilder()
        writeJml(sb, startTagOnly = true)
        sb.toString().hashCode()
    }

    override fun compareTo(other: JmlEvent): Int {
        if (truncatedTime != other.truncatedTime) {
            return truncatedTime.compareTo(other.truncatedTime)
        }
        if (juggler != other.juggler) {
            return juggler.compareTo(other.juggler)
        }
        if (hand != other.hand) {
            // ordering is so right hand sorts before left:
            return other.hand.compareTo(hand)
        }
        // shouldn't get here; shouldn't have multiple events for the same
        // juggler/hand at a single time
        return x.compareTo(other.x)
    }

    //--------------------------------------------------------------------------
    // Event transformations
    //--------------------------------------------------------------------------

    @Throws(JuggleExceptionUser::class)
    private fun withHandString(handString: String): JmlEvent {
        var newJuggler: Int
        var newHand: Int

        val index = handString.indexOf(":")
        if (index == -1) {
            newJuggler = 1
            newHand = if (handString.equals("left", ignoreCase = true)) {
                LEFT_HAND
            } else if (handString.equals("right", ignoreCase = true)) {
                RIGHT_HAND
            } else {
                val message = jlGetStringResource(Res.string.error_hand_name)
                throw JuggleExceptionUser("$message '$handString'")
            }
        } else {
            newJuggler = handString.take(index).toInt()
            val substr = handString.substring(index + 1)
            newHand = if (substr.equals("left", ignoreCase = true)) {
                LEFT_HAND
            } else if (substr.equals("right", ignoreCase = true)) {
                RIGHT_HAND
            } else {
                val message = jlGetStringResource(Res.string.error_hand_name)
                throw JuggleExceptionUser("$message '$handString'")
            }
        }
        return copy(juggler = newJuggler, hand = newHand)
    }


    fun withTransition(trans: JmlTransition): JmlEvent {
        return copy(transitions = transitions + trans)
    }

    @Suppress("unused")
    fun withoutTransitionAtIndex(index: Int): JmlEvent {
        return copy(
            transitions = transitions.filterIndexed { i, _ -> i != index }
        )
    }

    fun withoutTransition(trans: JmlTransition): JmlEvent {
        return copy(
            transitions = transitions.filter { it != trans }
        )
    }

    companion object {
        // hand descriptors
        const val LEFT_HAND: Int = 1
        const val RIGHT_HAND: Int = 2

        fun handIndex(handDescriptor: Int) = if (handDescriptor == LEFT_HAND) 0 else 1

        //----------------------------------------------------------------------
        // Constructing JmlEvents
        //----------------------------------------------------------------------

        @Throws(JuggleExceptionUser::class)
        fun fromJmlNode(
            current: JmlNode,
            numberOfJugglers: Int,
            numberOfPaths: Int,
            loadingJmlVersion: String = JmlDefs.CURRENT_JML_VERSION
        ): JmlEvent {
            var tempx = 0.0
            var tempy = 0.0
            var tempz = 0.0
            var tempt = 0.0
            var handstr: String? = null

            try {
                for ((name, value) in current.attributes.entries) {
                    if (name.equals("x", ignoreCase = true)) {
                        tempx = jlParseFiniteDouble(value)
                    } else if (name.equals("y", ignoreCase = true)) {
                        tempy = jlParseFiniteDouble(value)
                    } else if (name.equals("z", ignoreCase = true)) {
                        tempz = jlParseFiniteDouble(value)
                    } else if (name.equals("t", ignoreCase = true)) {
                        tempt = jlParseFiniteDouble(value)
                    } else if (name.equals("hand", ignoreCase = true)) {
                        handstr = value
                    }
                }
            } catch (_: NumberFormatException) {
                val message = jlGetStringResource(Res.string.error_event_coordinate)
                throw JuggleExceptionUser(message)
            }

            // JML version 1.0 used a different coordinate system -- convert
            if (loadingJmlVersion == "1.0") {
                tempy = tempz.also { tempz = tempy }
            }

            if (handstr == null) {
                val message = jlGetStringResource(Res.string.error_unspecified_hand)
                throw JuggleExceptionUser(message)
            }

            val result = JmlEvent(
                x = tempx,
                y = tempy,
                z = tempz,
                t = tempt
            ).withHandString(handstr)

            if (result.juggler !in 1..numberOfJugglers) {
                val message = jlGetStringResource(Res.string.error_juggler_out_of_range)
                throw JuggleExceptionUser(message)
            }

            val newTransitions = buildList {
                // process current event node children
                for (child in current.children) {
                    val childNodeType = child.nodeType
                    var childPath: String? = null
                    var childTranstype: String? = null
                    var childMod: String? = null

                    for ((name, value) in child.attributes.entries) {
                        if (name.equals("path", ignoreCase = true)) {
                            childPath = value
                        } else if (name.equals("type", ignoreCase = true)) {
                            childTranstype = value
                        } else if (name.equals("mod", ignoreCase = true)) {
                            childMod = value
                        }
                    }

                    if (childPath == null) {
                        val message = jlGetStringResource(Res.string.error_no_path)
                        throw JuggleExceptionUser(message)
                    }

                    val pathNum = childPath.toInt()
                    if (pathNum !in 1..numberOfPaths) {
                        val message = jlGetStringResource(Res.string.error_path_out_of_range)
                        throw JuggleExceptionUser(message)
                    }

                    if (childNodeType.equals("throw", ignoreCase = true)) {
                        add(
                            JmlTransition(
                                type = JmlTransition.TRANS_THROW,
                                path = pathNum,
                                throwType = childTranstype,
                                throwMod = childMod
                            )
                        )
                    } else if (childNodeType.equals("catch", ignoreCase = true) &&
                        childTranstype.equals("soft", ignoreCase = true)
                    ) {
                        add(
                            JmlTransition(
                                type = JmlTransition.TRANS_SOFTCATCH,
                                path = pathNum
                            )
                        )
                    } else if (childNodeType.equals("catch", ignoreCase = true) &&
                        childTranstype.equals("grab", ignoreCase = true)
                    ) {
                        add(
                            JmlTransition(
                                type = JmlTransition.TRANS_GRABCATCH,
                                path = pathNum
                            )
                        )
                    } else if (childNodeType.equals("catch", ignoreCase = true)) {
                        add(JmlTransition(JmlTransition.TRANS_CATCH, pathNum, null, null))
                    } else if (childNodeType.equals("holding", ignoreCase = true)) {
                        add(
                            JmlTransition(
                                type = JmlTransition.TRANS_HOLDING,
                                path = pathNum
                            )
                        )
                    }

                    if (child.children.isNotEmpty()) {
                        val message = jlGetStringResource(Res.string.error_event_subtag)
                        throw JuggleExceptionUser(message)
                    }
                }
            }

            return result.copy(transitions = newTransitions)
        }
    }
}